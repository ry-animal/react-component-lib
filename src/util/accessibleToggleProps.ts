export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
