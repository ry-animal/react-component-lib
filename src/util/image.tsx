import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
