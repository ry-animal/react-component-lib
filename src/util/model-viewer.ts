/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
