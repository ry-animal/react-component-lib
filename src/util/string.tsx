const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
