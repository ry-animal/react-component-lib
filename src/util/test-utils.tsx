import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
