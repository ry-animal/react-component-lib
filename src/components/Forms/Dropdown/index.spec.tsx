import userEvent from "@testing-library/user-event";
import { Dropdown } from ".";
import { render, screen } from "../../../util/test-utils";

describe("<Dropdown />", () => {
  const sampleProps = () => ({
    label: "Label",
    options: ["Option 1", "Option 2"],
    onSelectItem: jest.fn(),
  });

  describe("rendering", () => {
    test("should render dropdown and matches snapshot", () => {
      expect(
        render(<Dropdown {...sampleProps()} />).asFragment()
      ).toMatchSnapshot();
    });

    test("should render dropdown and matches snapshot (with custom options)", () => {
      const customOptions = [
        <h1 key="1" data-testid="h1-custom">
          Custom 1
        </h1>,
        <h2 key="2">Custom 2</h2>,
      ];
      expect(
        render(
          <Dropdown {...sampleProps()} options={customOptions} />
        ).asFragment()
      ).toMatchSnapshot();
      const toggle = screen.getByRole("button");
      userEvent.click(toggle);
      const custom = screen.getByTestId("h1-custom");
      expect(custom).toBeTruthy();
    });

    test("should render dropdown and matches snapshot (disabled)", () => {
      expect(
        render(<Dropdown {...sampleProps()} disabled />).asFragment()
      ).toMatchSnapshot();
    });
  });

  describe("event handling", () => {
    // Default is only present when dropdown not requried
    test("onSelectItem is called when default selected", () => {
      const props = sampleProps();
      render(<Dropdown {...props} />);
      const toggle = screen.getByRole("button");
      userEvent.click(toggle);
      const [option] = screen.getAllByRole("option");
      userEvent.click(option);
      expect(props.onSelectItem).toHaveBeenCalled();
    });

    test("selecting option closes menu and updates option", () => {
      const props = sampleProps();
      render(<Dropdown {...props} />);
      const toggle = screen.getByRole("button");
      userEvent.click(toggle);
      const [option] = screen.getAllByRole("option");
      userEvent.click(option);

      expect(toggle.textContent).toBe(props.options[0]);
    });

    test("pressing clear button resets selected item", () => {
      const props = sampleProps();
      render(<Dropdown {...props} />);
      const toggle = screen.getByRole("button");
      userEvent.click(toggle);
      const [option] = screen.getAllByRole("option");
      userEvent.click(option);

      const clearButton = screen.getByLabelText("clear selection");
      userEvent.click(clearButton);

      expect(toggle.textContent).toBe("Select One");
    });
  });
});
