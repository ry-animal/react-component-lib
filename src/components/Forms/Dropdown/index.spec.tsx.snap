// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> rendering should render dropdown and matches snapshot (disabled) 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: 'Poppins',sans-serif;
  gap: 12px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  -webkit-transition: box-shadow 0.3s ease-out;
  transition: box-shadow 0.3s ease-out;
  box-shadow: 0 0 0 1px #DBDBDB;
  border-radius: 4px;
}

.c2:hover {
  box-shadow: 0 0 0 2px #151515;
}

.c1 {
  color: #151515;
  margin: 0;
}

.c4 {
  position: absolute;
  z-index: 10;
  right: 15px;
  top: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
}

.c3 {
  background: #FFFFFF;
  color: #151515;
  border-radius: 4px;
  padding: 0 16px;
  font-weight: 600;
  font-size: 16px;
  height: 48px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  cursor: pointer;
  line-height: 24px;
  border: none;
  outline: none;
}

.c5 {
  display: none;
  padding: 8px;
  gap: 4px;
  z-index: 10;
  background: #FFFFFF;
  top: 56px;
  border-radius: 8px;
  box-shadow: 0px 2px 16px RGBA(0,0,0,0.16);
  width: 100%;
  box-sizing: border-box;
  list-style-type: none;
  border: none;
  outline: none;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  max-height: 400px;
  overflow-y: auto;
  margin: 0;
  position: absolute;
}

<div
    class="c0"
  >
    <label
      class="c1"
      for="downshift-3-toggle-button"
      id="downshift-3-label"
    >
      Label
    </label>
    <div
      class="c2"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-3-label downshift-3-toggle-button"
        class="c3"
        disabled=""
        id="downshift-3-toggle-button"
        type="button"
      >
        Select One
      </button>
      <div
        class="c4"
      >
        <svg
          fill="none"
          height="30"
          viewBox="0 0 30 30"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.5 11.25L15 18.75L22.5 11.25"
            stroke="#151515"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <ul
        aria-labelledby="downshift-3-label"
        class="c5"
        id="downshift-3-menu"
        role="listbox"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Dropdown /> rendering should render dropdown and matches snapshot (with custom options) 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: 'Poppins',sans-serif;
  gap: 12px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  -webkit-transition: box-shadow 0.3s ease-out;
  transition: box-shadow 0.3s ease-out;
  box-shadow: 0 0 0 1px #DBDBDB;
  border-radius: 4px;
}

.c2:hover {
  box-shadow: 0 0 0 2px #151515;
}

.c1 {
  color: #151515;
  margin: 0;
}

.c4 {
  position: absolute;
  z-index: 10;
  right: 15px;
  top: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
}

.c3 {
  background: #FFFFFF;
  color: #151515;
  border-radius: 4px;
  padding: 0 16px;
  font-weight: 600;
  font-size: 16px;
  height: 48px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  cursor: pointer;
  line-height: 24px;
  border: none;
  outline: none;
}

.c5 {
  display: none;
  padding: 8px;
  gap: 4px;
  z-index: 10;
  background: #FFFFFF;
  top: 56px;
  border-radius: 8px;
  box-shadow: 0px 2px 16px RGBA(0,0,0,0.16);
  width: 100%;
  box-sizing: border-box;
  list-style-type: none;
  border: none;
  outline: none;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  max-height: 400px;
  overflow-y: auto;
  margin: 0;
  position: absolute;
}

<div
    class="c0"
  >
    <label
      class="c1"
      for="downshift-1-toggle-button"
      id="downshift-1-label"
    >
      Label
    </label>
    <div
      class="c2"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-1-label downshift-1-toggle-button"
        class="c3"
        id="downshift-1-toggle-button"
        type="button"
      >
        Select One
      </button>
      <div
        class="c4"
      >
        <svg
          fill="none"
          height="30"
          viewBox="0 0 30 30"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.5 11.25L15 18.75L22.5 11.25"
            stroke="#151515"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <ul
        aria-labelledby="downshift-1-label"
        class="c5"
        id="downshift-1-menu"
        role="listbox"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Dropdown /> rendering should render dropdown and matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: 'Poppins',sans-serif;
  gap: 12px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  -webkit-transition: box-shadow 0.3s ease-out;
  transition: box-shadow 0.3s ease-out;
  box-shadow: 0 0 0 1px #DBDBDB;
  border-radius: 4px;
}

.c2:hover {
  box-shadow: 0 0 0 2px #151515;
}

.c1 {
  color: #151515;
  margin: 0;
}

.c4 {
  position: absolute;
  z-index: 10;
  right: 15px;
  top: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
}

.c3 {
  background: #FFFFFF;
  color: #151515;
  border-radius: 4px;
  padding: 0 16px;
  font-weight: 600;
  font-size: 16px;
  height: 48px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  cursor: pointer;
  line-height: 24px;
  border: none;
  outline: none;
}

.c5 {
  display: none;
  padding: 8px;
  gap: 4px;
  z-index: 10;
  background: #FFFFFF;
  top: 56px;
  border-radius: 8px;
  box-shadow: 0px 2px 16px RGBA(0,0,0,0.16);
  width: 100%;
  box-sizing: border-box;
  list-style-type: none;
  border: none;
  outline: none;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  max-height: 400px;
  overflow-y: auto;
  margin: 0;
  position: absolute;
}

<div
    class="c0"
  >
    <label
      class="c1"
      for="downshift-0-toggle-button"
      id="downshift-0-label"
    >
      Label
    </label>
    <div
      class="c2"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-0-label downshift-0-toggle-button"
        class="c3"
        id="downshift-0-toggle-button"
        type="button"
      >
        Select One
      </button>
      <div
        class="c4"
      >
        <svg
          fill="none"
          height="30"
          viewBox="0 0 30 30"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.5 11.25L15 18.75L22.5 11.25"
            stroke="#151515"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <ul
        aria-labelledby="downshift-0-label"
        class="c5"
        id="downshift-0-menu"
        role="listbox"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;
